pkg/operator/ceph/cluster/osd/osd_on_pvc_test.go

```diff
@@ -0,0 +1,720 @@
+			resources = append(resources, ro.resourceName)
+		}
+	}
+	return resources
+}
+
+func (r *resourceOperationList) OperationsInOrder() []string {
+	ops := []string{}
+	for _, ro := range r.List() {
+		ops = append(ops, ro.operation)
+	}
+	return ops
+}
+
+// The definition for this test is a wrapper for the test function that adds a timeout
+func TestOSDsOnPVC(t *testing.T) {
+	// This test runs in less than 350 milliseconds on a 6-core CPU w/ 16GB RAM
+	done := make(chan bool)
+	timeout := time.After(5 * 350 * time.Millisecond)
+
+	go func() {
+		// use defer because t.Fatal will kill this goroutine, and we always want done set if the
+		// func stops running
+		defer func() { done <- true }()
+		// run the actual test
+		testOSDsOnPVC(t)
+	}()
+
+	select {
+	case <-timeout:
+		t.Fatal("Test timed out. This is a test failure.")
+	case <-done:
+	}
+}
+
+// This is the actual test. If it hangs, we should consider that an error. Writing a timeout in
+// tests requires running the test in a goroutine, so there is a timeout wrapper above
+func testOSDsOnPVC(t *testing.T) {
+	oldLogger := *logger
+	defer func() { logger = &oldLogger }() // reset logger to default after this test
+	logger.SetLevel(capnslog.TRACE)        // want more log info for this test if it fails
+
+	oldOpportunisticDuration := osdOpportunisticUpdateDuration
+	defer func() { osdOpportunisticUpdateDuration = oldOpportunisticDuration }()
+	// lower the opportunistic update check duration for unit tests to speed them up
+	osdOpportunisticUpdateDuration = 1 * time.Millisecond
+
+	// make infof function for helping debugging by identifying logs output by the test itself
+	infof := func(format string, args ...interface{}) {
+		logger.Infof(t.Name()+": "+format, args...)
+	}
+
+	namespace := "osd-on-pvc"
+	ctx := context.TODO()
+	// we don't need to create actual nodes for this test, but this is the set of nodes which should
+	// we will use to create fake placements for OSD prepare job pods
+	nodeList := []string{"node1", "node2", "node3"}
+	pickNode := func() string {
+		// simply return the node names from the list in a round-robin fashion
+		node := nodeList[0]
+		nodeList = append(nodeList[1:], node)
+		return node
+	}
+	nextOSDID := 0
+	osdIDMap := map[string]int{}
+	getOSDID := func(configMapName string) int {
+		if id, ok := osdIDMap[configMapName]; ok {
+			infof("ConfigMap %q has existing OSD ID %d", configMapName, id)
+			return id
+		}
+		id := nextOSDID
+		osdIDMap[configMapName] = id
+		nextOSDID++
+		infof("Generated new OSD ID %d for ConfigMap %q", id, configMapName)
+		return id
+	}
+
+	// set up a fake k8s client set and watcher to generate events that the operator will listen to
+	clientset := fake.NewSimpleClientset()
+	d := clientset.Discovery()
+	fd, ok := d.(*fakediscovery.FakeDiscovery)
+	assert.True(t, ok)
+	// Set a fake k8s server version
+	fd.FakedServerVersion =
+		&version.Info{
+			Major:      "1",
+			Minor:      "13",
+			GitVersion: "v1.13.2",
```

Comment:
Why such an old version?

Reply:
It requires at least k8s v13 IIRC for some feature, so I used that in the test.